### PR TITLE
A proposal API for checking the input to be compatible with the model

### PR DESCRIPTION
**Pull request by BloodAxe**
This PR adds API for a mixin SupportsInputShapeCheck to validate the input size whether it is compatible with the model.

The goal of this PR is to demonstrate the basic concept where these size checks may happen (At the start of training, during export and during predict). I'm not sure whether we want to have these checks at each forward() call of the model. But I'm looking forward to hear your toughs on this.

Currently the following models are inheriting from SupportsInputShapeCheck and corresponding methods are implemented (Yet values are hard-coded and would not reflect changes in model configuration via arch_params that may change the strides configuration):

* `YoloNAS`
* `YoloNAS-Pose`
* `PPYoloE`
* `YoloX`
* All models inheriting from `SegmentationModule` and `STDC` model

### CODE REVIEW

File: src/super_gradients/module_interfaces/supports_input_shape_check.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    def get_minimum_input_shape_size(self) -> Optional[Tuple[int, ...]]:
+        """
+        Get the minimum input shape of the model.
+        Class should override this method if the model requires a minimum input shape.
+        Usually true for transformer-based models.
+        :return:
+        """
+        return None
+
+    def get_input_shape_steps(self) -> Optional[Tuple[int, ...]]:
```
Comment (Louis-Dupont) [inline src/super_gradients/module_interfaces/supports_input_shape_check.py:51]: `shape_steps` in `get_input_shape_steps` is not super intuitive name IMO.
I called it `shape_multiple` in the `AutoPadding`, I think it's a little clearer.
I also thought of `shape_modulo` but I think its a bit less clear as well. What do you think?

**Review by Louis-Dupont — APPROVED** [review]
Looks great, I'm good with it.

Still, I have a small note about the usage of Mixins which I believe is slowly increasing the complexity of the code:
- Some models inherit some mixins but not all it's not always easy to keep track
- Adding a new mixin is adding new code to each class that wants to inherit it. Here for instance, we needs to add 3 method to the models, on top of all the methods already required to these models. This leads to the last point:
- I believe a proper way of scaling features would be more through composition with classes responsible for these features, not inheritance like we do with all the mixins.

I had this feeling a couple of days ago, when navigating between Mixins. I still think these are a great solution for the problems we are facing, especially to keep backward compatibility, and I don't think it's a problem to introduce a new one here. 
I'm just raising a flag that we are starting to accumulate these and I'm wondering if there is a way to improve our current approach.

File: src/super_gradients/module_interfaces/supports_input_shape_check.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/module_interfaces/supports_input_shape_check.py:51]: The name of the method for sure can be improved. I had hard time formulating it's name )

[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/module_interfaces/supports_input_shape_check.py:51]: > Adding a new mixin is adding new code to each class that wants to inherit it. Here for instance, we needs to add 3 method to the models, on top of all the methods already required to these models. 

The idea of this mixin was to have the clear interface. The default implementation does not require one to implement any of these methods. So one can derive `SgModule(..., SupportsInputShapeCheck) and it would not break existing code. However down the class hierarcy, one can implement some methods to validate input shape. 

> I believe a proper way of scaling features would be more through composition with classes responsible for these features, not inheritance like we do with all the mixins.

Can you please show an example in pseud-code what you mean? I'm not sure I'm getting it.

[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply src/super_gradients/module_interfaces/supports_input_shape_check.py:51]: > The idea of this mixin was to have the clear interface. The default implementation does not require one to implement any of these methods. So one can derive `SgModule(..., SupportsInputShapeCheck) and it would not break existing code. However down the class hierarcy, one can implement some methods to validate input shape.

Sure I'm aware of that, and I believe it's overall a great way of doing it. I was just raising the downsides, which are that it eventually leads to very "flat" classes with lower cohesion + when someone wants to implement a new model, he needs to remember all the mixin that is available if he wants to support these features.


> Can you please show an example in pseud-code what you mean? I'm not sure I'm getting it.

Well generally speaking I don't know, and that's the problem 😅 

In this specific case, I see something we could maybe improve but it would not remove the need for mixins.

``` python
class InputValidator:
    def __init__(self, exact_image_size, min_image_size, multiple_shape):
        ...

    def validate(self, image):
        ...


class SupportsInputShapeCheck(ABC):
    @staticmethod
    @abstractmethod
    def validate_image(image):
        ...


class MyModel(SupportsInputShapeCheck):
    @staticmethod
    def validate_image(image):
        return InputValidator(exact_image_size=..., min_image_size=..., multiple_shape=...).validate(image)
        # return validate_image(exact_image_size=..., min_image_size=..., multiple_shape=..., image=image) # OR as a function directly

MyModel.validate_image(image)
```

Basically, the logic is moved outside of the Model.

But as you can see, this would still require the Mixin to declare which function support this feature.
So it does reduce the amount of methods to implement per class, and IMO leads to lower code complexity, but it does not remove the need for the inheritance of mixin.

**Review by ofrimasad — DISMISSED** [review]
LGTM

**Review by Louis-Dupont — DISMISSED** [review]
LGTM
